macos/FluentUI/Button.swift

```diff
@@ -0,0 +1,188 @@
+}
+
+@objc public enum ButtonStyle: Int, CaseIterable {
+    case primaryFilled
+    case primaryOutline
+    case borderless
+}
+
+// MARK: - Button
+
+/// A fluent styled button, with hover effects and a corner radius.
+@IBDesignable
+@objc(MSFButton)
+open class Button: NSButton {
+	/// The primary color of the button, AKA, the fill color in the primaryFilled style, and the outline in the primaryOutline style
+	@objc open var primaryColor: NSColor = NSColor.clear
+	
+	private var trackingArea: NSTrackingArea?
+	
+	private var mouseEntered: Bool = false
+	private var mouseClicked: Bool = false
+	
+	override open var wantsUpdateLayer: Bool {
+		return true
+	}
+
+	private let style: ButtonStyle
+	
+	@objc public init(title: String, style: ButtonStyle = .primaryFilled) {
+		self.style = style
+        super.init(frame: .zero)
+		self.title = title
+        initialize()
+    }
+	
+    public override init(frame: CGRect) {
+		preconditionFailure()
+    }
+
+    public required init?(coder aDecoder: NSCoder) {
+		preconditionFailure()
+    }
+	
+	open func initialize() {
+		// Do common initialization work
+		isBordered = false
+		wantsLayer = true
+
+		if #available(macOS 10.14, *) {
+			primaryColor = NSColor.controlAccentColor
+		} else {
+			primaryColor = NSColor.systemBlue
+		}
+	}
+	
+	override public func updateTrackingAreas() {
+		super.updateTrackingAreas()
+		
+		// Remove existing one
+		if let trackingArea = trackingArea {
+			removeTrackingArea(trackingArea)
+			self.trackingArea = nil
+		}
+		
+		// Create a new one
+		let opts: NSTrackingArea.Options = [.mouseEnteredAndExited, .activeAlways]
+		let trackingArea = NSTrackingArea(
+			rect: bounds,
+			options: opts,
+			owner: self,
+			userInfo: nil
+		)
+		addTrackingArea(trackingArea)
+		self.trackingArea = trackingArea
+	}
+	
+	override public func mouseEntered(with event: NSEvent) {
+		mouseEntered = true
+		needsDisplay = true
+	}
+
+	override public func mouseExited(with event: NSEvent) {
+		mouseEntered = false
+		needsDisplay = true
+	}
+	
+	override public func mouseDown(with event: NSEvent) {
+		mouseClicked = true
+		needsDisplay = true
+	}
+	
+	override public func mouseUp(with event: NSEvent) {
+		mouseClicked = false
+		needsDisplay = true
+	}
+	
+	open override func updateLayer() {
+		layer?.borderWidth = Constants.borderWidth
+		layer?.masksToBounds = true
+		layer?.cornerRadius = Constants.cornerRadius
+				
+		var fillColor : NSColor = NSColor.clear
+		var outlineColor : NSColor = NSColor.clear
+		var textColor : NSColor = NSColor.white
+		
+		switch style {
+		case .primaryFilled:
+			fillColor = primaryColor
+			break
+		case .primaryOutline:
+			outlineColor = primaryColor.withAlphaComponent(0.4)
+			textColor = primaryColor
+			break
+		case .borderless:
+			textColor = primaryColor
+			break
+		}
+		
+		if (isEnabled) {
+			if (mouseClicked) {
+				if #available(macOS 10.14, *) {
+					layer?.backgroundColor = fillColor.withSystemEffect(.pressed).cgColor
```

Comment:
Side note: I didn't know these "withSystemEffect" API's existed till today and I'm glad they do 😁